src/main/scala/chisel3/experimental/inlinetest/InlineTest.scala

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.experimental.inlinetest
+
+import chisel3._
+import chisel3.experimental.hierarchy.{Definition, Instance}
+
+/** Per-test parametrization needed to build a testharness that instantiates
+  * the DUT and elaborates a test body.
+  *
+  *  @tparam M the type of the DUT module
+  *  @tparam R the type of the result returned by the test body
+  */
+class TestParameters[M <: RawModule, R] private[inlinetest] (
+  /** The [[desiredName]] of the DUT module. */
+  val dutName: String,
+  /** The user-provided name of the test. */
+  val testName: String,
+  /** A Definition of the DUT module. */
+  val dutDefinition: Definition[M],
+  /** The body for this test, returns a result. */
+  val testBody: Instance[M] => R
+) {
+  def desiredTestModuleName = s"test_${dutName}_${testName}"
+}
+
+/** TestHarnesses for inline tests without clock and reset IOs should extend this. This
+  * trait sets the correct desiredName for the module, instantiates the DUT, and provides
+  * methods to elaborate the test.
+  *
+  *  @tparam M the type of the DUT module
+  *  @tparam R the type of the result returned by the test body
+  */
+trait TestHarnessRawModule[M <: RawModule, R] extends Public { this: RawModule =>
+  def test: TestParameters[M, R]
+  override def desiredName = test.desiredTestModuleName
+  val dut = Instance(test.dutDefinition)
+  final def elaborateTest(): R = test.testBody(dut)
+}
+
+/** TestHarnesses for inline tests should extend this. This trait sets the correct desiredName for
+ *  the module, instantiates the DUT, and provides methods to elaborate the test. By default, the
+ *  reset is synchronous, but this can be changed by overriding [[resetType]].
+ *
+ *  @tparam M the type of the DUT module
+ *  @tparam R the type of the result returned by the test body
+ */
+trait TestHarnessModule[M <: RawModule, R] extends TestHarnessRawModule[M, R] { this: Module =>
+  override def resetType = Module.ResetType.Synchronous
+}
+
+/** An implementation of a testharness generator.
+  *
+  *  @tparam M the type of the DUT module
+  *  @tparam R the type of the result returned by the test body
+  */
+trait TestHarnessGenerator[M <: RawModule, R] {
```

Comment:
Can the PR update this comment to explicitly call out that this is a type class.